### PR TITLE
Adding validator to check the spec

### PR DESCRIPTION
**Pull request by bluemonk3y**
subtask part of #86 - where the spec should be validated against the current env . This is the first pass. later will include cover missinmg functionality i.e. compatible topic config changes, schema validation etc

### CODE REVIEW

File: kafka/src/main/java/io/specmesh/kafka/ProvisionStatus.java [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+@AllArgsConstructor(access = AccessLevel.PRIVATE)
+@NoArgsConstructor(access = AccessLevel.PRIVATE)
+@Accessors(fluent = true)
+@SuppressFBWarnings
+public class ProvisionStatus {
+    private Topics topics;
+    private Schemas schemas;
+    private Acls acls;
+
+    /** Process the accumulated state */
```
Comment (big-andy-coates) [inline kafka/src/main/java/io/specmesh/kafka/ProvisionStatus.java:49]: Bogus javadoc

File: kafka/src/main/java/io/specmesh/kafka/ProvisionStatus.java [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+    /** Operation result */
+    public enum CRUD {
+        CREATE,
+        CREATED,
+        READ,
+        UPDATE,
+        DELETE,
+        IGNORED,
+        FAILED
+    }
```
Comment (big-andy-coates) [inline kafka/src/main/java/io/specmesh/kafka/ProvisionStatus.java]: lol - clearly not `CRUD` as its more than C,R,U,D.

Java docs required to explain the difference between `CREATE` and `CREATED` for sure!

File: kafka/src/main/java/io/specmesh/kafka/ProvisionStatus.java [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+                                        if (wasCreated.isPresent()) {
+                                            aclStatus.crud(CRUD.CREATED);
+                                        }
+                                        if (exception != null) {
+                                            aclStatus.crud(CRUD.FAILED);
+                                            aclStatus.exception(exception);
+                                        }
+                                        return aclStatus.build();
+                                    })
+                            .collect(Collectors.toMap(AclStatus::name, aclStatus -> aclStatus));
```
Comment (big-andy-coates) [inline kafka/src/main/java/io/specmesh/kafka/ProvisionStatus.java:245]: Yuk - not reviewing

File: kafka/src/main/java/io/specmesh/kafka/Provisioner.java [hunk 4]
```diff
@@ -209,17 +257,22 @@ private static String readSchemaContent(final Path schemaPath) {
...
+     * @param validateMode for mode of operation
      * @param apiSpec the api spec.
      * @param adminClient th admin client for the cluster.
      * @return status of provisioning
      * @throws ProvisioningException on interrupt
      */
-    public static AclStatus provisionAcls(final KafkaApiSpec apiSpec, final Admin adminClient) {
-        final var aclStatus = AclStatus.builder();
+    public static ProvisionStatus.Acls provisionAcls(
+            final boolean validateMode, final KafkaApiSpec apiSpec, final Admin adminClient) {
```
Comment (big-andy-coates) [inline kafka/src/main/java/io/specmesh/kafka/Provisioner.java:267]: As above, I'd make this private and use the builder style API

File: kafka/src/main/java/io/specmesh/kafka/ProvisionStatus.java (hunk 2 above)
Comment (bluemonk3y) [reply kafka/src/main/java/io/specmesh/kafka/ProvisionStatus.java]: yeah - added the jdocs and renamed it to status

File: kafka/src/main/java/io/specmesh/kafka/Provisioner.java [hunk 5]
```diff
@@ -160,23 +193,35 @@ public static List<SchemaStatus> provisionSchemas(
...
+                                        topic,
+                                        schemaSubject,
+                                        schemaRef,
+                                        schemaPath,
+                                        readSchemaContent(schemaPath));
+                        status.id(id);
+                        if (id == -1) {
+                            status.crud(CREATE);
+                        } else {
+                            status.crud(CREATED);
```
[the remark this answers is not in the document]
Comment (bluemonk3y) [reply kafka/src/main/java/io/specmesh/kafka/Provisioner.java:210]: the id will be returned when it is successfully registered...

File: kafka/src/main/java/io/specmesh/kafka/ProvisionStatus.java (hunk 1 above)
Comment (bluemonk3y) [reply kafka/src/main/java/io/specmesh/kafka/ProvisionStatus.java:49]: ill do the builder stuff in the next PR

File: kafka/src/main/java/io/specmesh/kafka/ProvisionStatus.java [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+        @Builder.Default private List<NewTopic> domainTopics = Collections.emptyList();
+        @Builder.Default private List<String> existingTopics = Collections.emptyList();
+        @Builder.Default private List<NewTopic> createTopics = Collections.emptyList();
+
+        @Builder.Default private List<NewTopic> createdTopics = Collections.emptyList();
+
+        private Exception exception;
+        private Map<String, TopicStatus> status;
+
+        /** Process the accumulated state */
```
[the remark this answers is not in the document]
Comment (bluemonk3y) [reply kafka/src/main/java/io/specmesh/kafka/ProvisionStatus.java:74]: ill do the builder stuff in the next PR

File: kafka/src/main/java/io/specmesh/kafka/ProvisionStatus.java (hunk 3 above)
Comment (bluemonk3y) [reply kafka/src/main/java/io/specmesh/kafka/ProvisionStatus.java:245]: ill do the builder stuff and sort this in the next PR

File: kafka/src/main/java/io/specmesh/kafka/Provisioner.java [hunk 7]
```diff
@@ -133,12 +164,14 @@ private static List<String> existingTopics(final Admin adminClient) {
...
      *
+     * @param validateMode for mode of operation
      * @param apiSpec the api spec
      * @param baseResourcePath the path under which external schemas are stored.
      * @param schemaRegistryClient the client for the schema registry
      * @return status of actions
      */
-    public static List<SchemaStatus> provisionSchemas(
+    public static Schemas provisionSchemas(
+            final boolean validateMode,
```
[the remark this answers is not in the document]
Comment (bluemonk3y) [reply kafka/src/main/java/io/specmesh/kafka/Provisioner.java:174]: ill do the builder stuff in the next PR

File: kafka/src/main/java/io/specmesh/kafka/Provisioner.java [hunk 8]
```diff
@@ -160,23 +193,35 @@ public static List<SchemaStatus> provisionSchemas(
...
+                        }
+
                     } catch (ProvisioningException ex) {
+                        status.crud(FAILED);
                         status.exception(ex);
                     }
                     statusList.add(status.build());
                 }));
-        return statusList;
+        return Schemas.builder().schemas(statusList).build();
```
[the remark this answers is not in the document]
Comment (bluemonk3y) [reply kafka/src/main/java/io/specmesh/kafka/Provisioner.java:219]: ill do the builder stuff in the next PR

File: kafka/src/main/java/io/specmesh/kafka/Provisioner.java (hunk 4 above)
Comment (bluemonk3y) [reply kafka/src/main/java/io/specmesh/kafka/Provisioner.java:267]: ill do the builder stuff in the next PR

File: kafka/src/test/java/io/specmesh/kafka/ProvisionerFunctionalTest.java [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+                            .count(),
+                    is(2L));
+            assertThat(
+                    schemaStatusMap.keySet(),
+                    is(
+                            containsInAnyOrder(
+                                    "simple.schema_demo._public.user_signed_up-value",
+                                    "simple.schema_demo._public.user_checkout-value")));
+        }
+    }
```
[the remark this answers is not in the document]
Comment (bluemonk3y) [reply kafka/src/test/java/io/specmesh/kafka/ProvisionerFunctionalTest.java:182]: yeah - agree - ill break out the builder stuff in the next PR - lets close this one off

File: kafka/src/main/java/io/specmesh/kafka/ProvisionStatus.java [hunk 10]
```diff
@@ -0,0 +1,252 @@
...
+                                                .replication(domainTopic.replicationFactor());
+                                        if (exception != null) {
+                                            status.crud(CRUD.FAILED);
+                                            status.exception(exception);
+                                        }
+                                        return status.build();
+                                    })
+                            .collect(Collectors.toMap(TopicStatus::name, item -> item));
+        }
+    }
```
[the remark this answers is not in the document]
Comment (bluemonk3y) [reply kafka/src/main/java/io/specmesh/kafka/ProvisionStatus.java:128]: ill do the builder stuff in the next PR

File: kafka/src/main/java/io/specmesh/kafka/ProvisionStatus.java (hunk 1 above)
Comment (bluemonk3y) [reply kafka/src/main/java/io/specmesh/kafka/ProvisionStatus.java:49]: ill do the builder stuff in the next PR

Comment (bluemonk3y) [reply kafka/src/main/java/io/specmesh/kafka/ProvisionStatus.java:49]: ill do the builder stuff in the next PR

File: kafka/src/main/java/io/specmesh/kafka/ProvisionStatus.java [hunk 11]
```diff
@@ -0,0 +1,252 @@
...
+    @Accessors(fluent = true)
+    @SuppressFBWarnings
+    public static class TopicStatus {
+        private String name;
+        private CRUD crud;
+        private int partitions;
+        private short replication;
+        private Map<String, String> configs;
+        private Exception exception;
+    }
```
[the remark this answers is not in the document]
Comment (bluemonk3y) [reply kafka/src/main/java/io/specmesh/kafka/ProvisionStatus.java:143]: ill do the builder stuff in the next PR

File: kafka/src/main/java/io/specmesh/kafka/ProvisionStatus.java [hunk 12]
```diff
@@ -0,0 +1,252 @@
...
+    public static class Schemas {
+        @Builder.Default private List<SchemaStatus> schemas = Collections.emptyList();
+        private Map<String, SchemaStatus> status;
+
+        public void build() {
+            status =
+                    schemas.stream()
+                            .collect(Collectors.toMap(SchemaStatus::schemaSubject, sss -> sss));
+        }
+    }
```
[the remark this answers is not in the document]
Comment (bluemonk3y) [reply kafka/src/main/java/io/specmesh/kafka/ProvisionStatus.java]: ill do the builder stuff in the next PR

**Pull request merged**